### PR TITLE
Deploy the site with a version selector

### PR DESCRIPTION
**Pull request by oprypin**
Preview: https://oprypin.github.io/crystal-book/1.0/ + https://oprypin.github.io/crystal-book/versions.json

This changes the approach to deploying the book:
Now the `master` branch will *not* be deployed, whereas a branch looking like `1.0` etc. will be deployed to https://crystal-lang.org/reference/1.0/

Note that I'm not distinguishing patch releases, that's very intentional (and I'd probably recommend that for API docs actually, but that's aside...)

The new approach resolves the problem that I had [mentioned](https://github.com/crystal-lang/crystal-book/pull/514#issuecomment-852262581) where we shouldn't push documentation changes before they actually make it into a release.
Well, now it's not a problem, people can just contribute to `master` branch, and then when the release comes, we only need to create a new release branch, and the new version will be published fully automatically.
Contributing directly to a branch is also possible; we can ask contributors to do that instead if we want some contribution to show up immediately, and later merge the release branch back into `master`.

This relies on this version selector feature https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning
The version selector feature is based on a "version manager" for MkDocs ["mike"](https://github.com/jimporter/mike) but we're actually not using it because it only supports pushing to Git whereas we're pushing to Amazon. So instead the same version file is being created directly by enumerating the available branches.

The popup ["you're not viewing the latest version"](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning) cannot be used from upstream because it's proprietary. Maybe eventually it will become public. I was unable to replicate that exact approach, so instead I just add a small label beside the version selector using a separate script.
image

The only thing is that I'm unsure at all how to integrate this with Amazon. I think we need to create `/latest/foo` and `/foo` redirects to `/1.1/foo`, like we're doing for API docs. I have not accomplished that part here and will need a consultation.

Deployment-specific code is commented out but of course I will put it back in when it's ready.

### CODE REVIEW

**Comment by straight-shoota** [comment]
> Now the master branch will not be deployed

I think we should probably still deploy master branch as development version of the book, corresponding to the master API docs.

This change introduces new URIs and we might want to consider if `crystal-lang.org/reference/x.y/` is the path where we want to mount them. We can redirect the old URIs in a backwards-compatible way, so it's maybe not necessary to discuss this together with this change.

File: .github/workflows/deploy-book.yml [hunk 1]
```diff
@@ -25,14 +26,56 @@ jobs:
...
-          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
-          aws-region: us-east-1
-      - name: Deploy book
-        if: github.event_name == 'push' && github.ref == 'refs/heads/master'
+          name: site
+          path: site
+
+  deploy:
+    name: Deploy book
+    if: github.event_name == 'push' && startsWith(github.ref, 'refs/heads/1.')
```
[the remark this answers is not in the document]
Comment (oprypin) [reply .github/workflows/deploy-book.yml:37]: If I knew how, I would've written it as such :/

File: .github/workflows/deploy-book.yml [hunk 2]
```diff
@@ -25,14 +26,56 @@ jobs:
...
+      - name: Build versions file
+        run: |
+          printf '[' >versions.json
+          delim=''; latest='"latest"'
+          for ver in $(git ls-remote --heads https://github.com/${{ github.repository }} | grep -P -o '(?<=refs/heads/)[0-9][0-9.]+' | sort -V -r); do
+            printf '%s{"version": "%s", "title": "%s", "aliases": [%s]}' "$delim" "$ver" "$ver" "$latest" >>versions.json
+            delim=', '; latest=''
+          done
+          printf ']' >>versions.json
+          test -n "$delim"  # Check that we wrote at least one version
```
[the remark this answers is not in the document]
Comment (oprypin) [reply .github/workflows/deploy-book.yml:55]: > I'd suggest to extract these commands to an external script

OK

> And can we query local branches instead?

OK. But it comes at the price of downloading the whole repo every time (versus not downloading anything at all)

**Comment by oprypin** [comment]
> I think we should probably still deploy master branch as development version of the book

OK

> We can redirect the old URIs in a backwards-compatible way, so it's maybe not necessary to discuss this together with this change.

Well I don't know that it is possible, that why I'm asking

File: .github/workflows/deploy-book.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (straight-shoota) [reply .github/workflows/deploy-book.yml:55]: Maybe it could be a configurable option to the script? Just spinning ideas...

**Review by straight-shoota — APPROVED** [review]

**Comment by straight-shoota** [comment]
@oprypin Can you resolve the conflict? I'd like to merge this.

File: scripts/docs-versions.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env sh
+set -e
+
+printf '['
+latest='"latest"'
+for ver in $(git ls-remote --heads "${1:-.}" | grep -P -o '(?<=refs/heads/)[0-9][0-9.]+' | sort -V -r); do
```
Comment (j8r) [inline scripts/docs-versions.sh:6]: There is the `--sort` option of `git ls-remote` that can be used.
It allows to treat tags as versions and sort them accordingly.

Comment (oprypin) [reply scripts/docs-versions.sh:6]: I think I will not change this. I trust the current approach a lot more than this:

```suggestion
for ver in $(git ls-remote --heads --sort='-version:refname' "${1:-.}" '[0-9].*' | grep -P -o '(?<=refs/heads/)[0-9.]+'); do
```

The sorting feature is quite finicky and I don't quite understand it.

I also wonder whether I intentionally avoided it the last time I did this.

**Comment by oprypin** [comment]
This is ready.
I thought maybe I'd also change the version warning to [the (now generally available) upstream feature](https://squidfunk.github.io/mkdocs-material/setup/setting-up-versioning/#version-warning) **but I decided to do that separately**.

**Comment by straight-shoota** [comment]
I've created a branch to test the CI workflow: https://github.com/crystal-lang/crystal-book/tree/1.2

It's live here: https://crystal-lang.org/reference/1.2/
Redirects are also in place: https://crystal-lang.org/reference/latest/

I think that means we can merge this branch into master. Then we can remove all the files without a version prefix from the bucket (except `versions.json`).
We still need to establish a mechanism for updating the AWS configuration to point the redirects to the latest version.

**Comment by oprypin** [comment]
>  We still need to establish a mechanism for updating the AWS configuration to point the redirects to the latest version.

Can we do that separately? It seems not essential for the moment, exactly because it's also possible to set up manually.

**Comment by straight-shoota** [comment]
Yes, definitely!

**Pull request merged**